include/picongpu/particles/Synchrotron/AlgorithmSynchrotron.hpp

```diff
@@ -0,0 +1,428 @@
+                    /// see Extended particle-in-cell schemes for physics in ultrastrong laser fields: Review and
+                    /// developments, A. Gonoskov et.Al.
+
+                    // delta
+                    float_X const r1r1 = randNr1 * randNr1;
+                    float_X const delta = r1r1 * randNr1;
+
+                    // calculate Heff
+                    float_X const mass = attribute::getMass(parentElectron[weighting_], parentElectron);
+                    float3_X const vel = Velocity()(parentElectron[momentum_], mass);
+
+                    float_X const Vmag = pmacc::math::l2norm(vel);
+                    float3_X const crossVB = pmacc::math::cross(vel, bField);
+                    float3_X const Vnorm = vel / Vmag; // Velocity normalized -> not L2Norm
+                    float_X const dotVnormE = pmacc::math::dot(Vnorm, eField);
+                    float3_X const eFieldPlusCrossVB = eField + crossVB;
+
+                    float_X HeffValue = pmacc::math::dot(eFieldPlusCrossVB, eFieldPlusCrossVB) - dotVnormE * dotVnormE;
+                    if(HeffValue <= 0)
+                    {
+                        return 0;
+                    }
+                    HeffValue = math::sqrt(HeffValue);
+
+
+                    // Calculate chi
+                    float_X const gamma = Gamma()(parentElectron[momentum_], mass);
+                    float_X const chi = -ELECTRON_CHARGE * HBAR * HeffValue * gamma
+                        / (ELECTRON_MASS * ELECTRON_MASS * SPEED_OF_LIGHT * SPEED_OF_LIGHT * SPEED_OF_LIGHT);
+
+                    // zq
+                    float_64 oneMinusDeltaOverDelta = (1 - delta) / delta;
+                    float_X zq = 2. / (3 * chi) / oneMinusDeltaOverDelta;
+
+                    // zq convert to index and F1 and F2
+                    const float_X zqExponent = math::log2(zq);
+                    const int16_t index
+                        = static_cast<int16_t>((zqExponent - minZqExponent) / stepWidthLogatihmicScale);
+
+                    float_X F1 = 0;
+                    float_X F2 = 0;
+                    float_X Ftemp = 0;
+                    if(index >= 0 && index < interpolationPoints - 2)
+                    {
+                        float_X zq1 = math::pow(2, minZqExponent + stepWidthLogatihmicScale * (index));
+                        float_X zq2 = math::pow(2, minZqExponent + stepWidthLogatihmicScale * (index + 1));
+                        float_X f = (zq - zq1) / (zq2 - zq1);
+
+                        /// @todo Test Logarithmic interpolation
+                        F1 = F1F2(index, 0);
+                        Ftemp = F1F2(index + 1, 0);
+                        F1 = math::pow(F1, 1 - f) * math::pow(Ftemp, f); // F1 = F1**((1-f)) * Ftemp**f
+
+                        F2 = F1F2(index, 1);
+                        Ftemp = F1F2(index + 1, 1);
+                        F2 = math::pow(F2, 1 - f) * math::pow(Ftemp, f); // F2 = F2**((1-f)) * Ftemp**f
```

Comment:
This reconstruction from the logarithmic interpolation table looks like it could be abstracted with an inline function and thus avoid code duplication.

Reply:
That would also help a reader of your code to quickly identify the code as an interpolation and not as some port of the probability calculation.

Reply:
Done, Now I'll test it

Reply:
compiles
done